### PR TITLE
Construct/Deconstruct helpers for Device Attestation

### PR DESCRIPTION
**Pull request by mleisner**
#### Problem

Provide data access methods for Device Attestation Verifier/DA (PR #9617 )

#### Change overview

- Construct Attestation Elements TLV
- Deconstruct Attestation Elements TLV
- Common routine CopySpantoMutableSpan
- Added ExtractVIDFromX509Cert and ExtractAKIDFromX509Cert

#### Testing

Matter Unit Tests

### CODE REVIEW

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.cpp:35]: Convention would be:

```
    kCertificationDeclarationTagId = 1,
    kAttestationNonceTagId = 2,
    kTimestampTagId = 3,
    kFirmwareInfoTagId = 4,
    kTagLast,
    kNumberOfValidTags = kTagLast
```

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (andy31415) [inline src/credentials/DeviceAttestationConstructor.cpp:56]: nit: this may feel more concise:

```cpp
uint64_t tag = tlvReader.GetTag();
uint32_t currentDecodeTagId = TLV::TagNumFromTag(tag);
```

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (andy31415) [inline src/credentials/DeviceAttestationConstructor.cpp:103]: `if (!seenProfile)` please

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (andy31415) [inline src/credentials/DeviceAttestationConstructor.cpp:40]: std::vector is generally not ok in chip code that will run on embedded due to heap usage.

Comment (tcarmelveilleux) [reply src/credentials/DeviceAttestationConstructor.cpp:40]: If this was in a separate file pair that was in a build target only used on Commissioners, it would be OK, but right now I agree this could/should be avoided. A good way to avoid it would be to have a vendorReserved data iterator that can be called separately, against the same CSV, to iterate the vendorReserved data.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:43]: agreed

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:44]: yes, agreed

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:103]: agreed

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp]: done

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp]: done, removed currentDecodeTagId

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.cpp:106]: This is not mandatory according to spec.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.cpp:47]: Please use a `ContiguousBufferTLVReader`, which has the following really practical method `CHIP_ERROR ContiguousBufferTLVReader::GetByteView(ByteSpan & data)` --> avoids needing to trust the length of current tag, and validates the underlying buffer won't be overrun by the view. You can use it for  `certificationDeclaration`  and for all entries in vendorReservedArray

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.cpp:37]: This method does not allow extraction of firmware info. Should be added as a TODO and left as a refactor.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.cpp:82]: The method here does not address the fact that all TLV elements *must* be in tag-order. A simple way to do it would be to keep track of "last context tag number", and start it at -1. In each test case, add what the expected previous tag should be.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.cpp:110]: This requires the caller to have allocated large buffers with unknown possible size. This is why I recommend using `ContiguousBufferTLVReader::GetByteView`, which got added later than this PR got started.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.cpp:136]: This should also be CHIP_ERROR_INVALID_ARGUMENT

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.cpp:162]: This makes assumptions about vendor tag usage that are not substantiated by the spec. Would recommend adding a TODO to make this spec compliant. Since this is optional and not going to be emitted by common code in the short term, I don't want to block this PR on that particular TODO as long as an issue is created and its link put in the TODO here.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.cpp:38]: If firmware info is not present, what happens?

File: src/credentials/DeviceAttestationConstructor.h
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.h:38]: What happens if any of the buffers are too for the bytespans? What does it look like for firmwareInfo to be missing? How can you tell? What are the possible errors, in which conditions? Please fill this to make it usable without reverse engineering the implementation.

File: src/credentials/DeviceAttestationConstructor.h
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.h:56]: Please provide sufficient detail to make it usable without reverse engineering the impl.

File: src/credentials/DeviceAttestationVerifier.h
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationVerifier.h:65]: All error codes in range 500..1000 relate to internal validation of data. Suggest just putting all of them in the 500 range

File: src/credentials/examples/DeviceAttestationVerifierExample.cpp
Comment (tcarmelveilleux) [inline src/credentials/examples/DeviceAttestationVerifierExample.cpp:209]: Add TODO to refactor once final vendor-specific data tags is handled.

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp
Comment (tcarmelveilleux) [inline src/credentials/tests/TestDeviceAttestationCredentials.cpp:178]: Why is the signature changed? Are you no longer using the official test vector?

File: src/credentials/tests/TestDeviceAttestationConstruction.cpp
Comment (tcarmelveilleux) [inline src/credentials/tests/TestDeviceAttestationConstruction.cpp:98]: This is a round-trip test. It does not externally validate that construction is done properly. 

Need to add tests to cover for parser doing:

1- At least one TLV tag out of order during parsing
2- firmware info present/missing
3- Being able to generate the exact byte stream from the spec test vector, and parsing the same byte string from test vector, without going through round-tripping.

File: src/credentials/tests/TestDeviceAttestationConstruction.cpp
Comment (tcarmelveilleux) [inline src/credentials/tests/TestDeviceAttestationConstruction.cpp:119]: Split test into construction/deconstruction. The test you currently have can be kept as a "RoundTrip" test, but we do need separate test cases as well for contruction+args --> expected octet string and octet string --> expected args.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:82]: ok, didn't see that constraint.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:110]: I agree, this was discussed.   Will investigate this.

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp
Comment (mleisner) [reply src/credentials/tests/TestDeviceAttestationCredentials.cpp:178]: will look at, thanks.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:106]: ok, thanks

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:47]: will update

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:136]: will update

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:162]: will add TODOs

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:38]: nothing -- it is not mandatory according to the spec.

File: src/credentials/DeviceAttestationConstructor.h
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.h:38]: will update

File: src/credentials/DeviceAttestationConstructor.h
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.h:56]: will update

File: src/credentials/DeviceAttestationVerifier.h
Comment (mleisner) [reply src/credentials/DeviceAttestationVerifier.h:65]: will update

File: src/credentials/examples/DeviceAttestationVerifierExample.cpp
Comment (mleisner) [reply src/credentials/examples/DeviceAttestationVerifierExample.cpp:209]: will update

File: src/credentials/tests/TestDeviceAttestationConstruction.cpp
Comment (mleisner) [reply src/credentials/tests/TestDeviceAttestationConstruction.cpp:98]: will update

File: src/credentials/tests/TestDeviceAttestationConstruction.cpp
Comment (mleisner) [reply src/credentials/tests/TestDeviceAttestationConstruction.cpp:119]: will update

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp]: done

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp]: removed the vector and use a simple array.

In the future, we'll use a 
class DAVendorReserved *vendorReserved

replacing

bytespan, &vendorId, &profileNUm

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp]: Current patch makes sure tags are in order.

Read A.2.4, should we have two while loops (instead of one big while loop) -- first for the context tags, second for profile tags.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [inline src/credentials/DeviceAttestationConstructor.cpp:24]: Remove this header, it is not used

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (tcarmelveilleux) [reply src/credentials/DeviceAttestationConstructor.cpp:118]: We cannot assume anything: there may be duplicate vendorId/profileNum. Eventually each entry should be accessible, even if duplicated. That is why I suggested we eventually (step 2) move to have this be a class, and have an iteration API for the vendor tags after parsing.

File: src/credentials/tests/TestDeviceAttestationCredentials.cpp
Comment (tcarmelveilleux) [reply src/credentials/tests/TestDeviceAttestationCredentials.cpp:178]: OK, good to know. The payload is the same, so that makes sense just the signature is affected.

File: src/credentials/tests/TestDeviceAttestationConstruction.cpp
Comment (tcarmelveilleux) [inline src/credentials/tests/TestDeviceAttestationConstruction.cpp:80]:
```suggestion
        // Test for success with entirely valid arguments
        err = ConstructAttestationElements(ByteSpan(certificationDeclaration), ByteSpan(attestationNonce), timestamp, ByteSpan(),
```

File: src/credentials/tests/TestDeviceAttestationConstruction.cpp
Comment (tcarmelveilleux) [inline src/credentials/tests/TestDeviceAttestationConstruction.cpp:269]: Suggest making the firmware info vector match example from spec:

```
firmwareInformation FirmwareInformation ::= {
  certificationDeclarationVersionNumber 5,
  -- The firmwareDigests contain two values, for two separate components.
  firmwareDigests {
    {
      digestAlgorithm {
        id id-sha256
      },
      digestHash '00112233445566778899AABBCCDDEEFF00112233445566778899AABBCCDDEEFF'H
    },
    {
      digestAlgorithm {
        id id-sha256
      },
      digestHash '101112131415161718191A1B1C1D1E1F101112131415161718191A1B1C1D1E1F'H
    }
  }
}
----

The above example would yield the following DER-encoded octet string:

----
30678001 05A16230 2FA00B80 09608648 01650304 02018120 00112233 44556677
8899AABB CCDDEEFF 00112233 44556677 8899AABB CCDDEEFF 302FA00B 80096086
48016503 04020181 20101112 13141516 1718191A 1B1C1D1E 1F101112 13141516
1718191A 1B1C1D1E 1F
----

```

**Review by woody-apple — APPROVED** [review]
Approving pending @tcarmelveilleux

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:118]: So duplicate tags can happen on deconstruction.   

Should we cause an error on "add vendor specific data" for creating?

**Review by hawk248 — APPROVED** [review]

File: src/credentials/tests/TestDeviceAttestationConstruction.cpp
Comment (mleisner) [reply src/credentials/tests/TestDeviceAttestationConstruction.cpp]: done

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp]: done

File: src/credentials/tests/TestDeviceAttestationConstruction.cpp
Comment (mleisner) [reply src/credentials/tests/TestDeviceAttestationConstruction.cpp:81]: done

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:117]: careful reading of the spec indicates there's nothing written having duplicate tags.

File: src/credentials/DeviceAttestationConstructor.cpp
Comment (mleisner) [reply src/credentials/DeviceAttestationConstructor.cpp:177]: Add issue #9825

**Review by tcarmelveilleux — APPROVED** [review]

**Pull request merged**
